### PR TITLE
Add iks-life-preserver terraform code

### PR DESCRIPTION
**Pull request by jpapejr**
Internal design doc and background - https://ibm.ent.box.com/notes/729730770506

tl;dr - this PR adds in a convenience Terraform configuration for simplifying the process for securely requesting debug assistance from another person (I.e. an IBM Support rep or developer) for the purposes of debugging an IKS cluster issue. This Terraform content will generate a Service ID and an associated API Key that can be used to authenticate to IBM Cloud via the CLI and only have Viewer/Reader access to a specific IKS cluster in a specific account. The user cannot update in-cluster RBAC or Secrets. To revoke the access, simply delete the IBM Cloud Schematics workspace and associated resources.

### CODE REVIEW

File: iks-life-preserver/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# iks-life-preserver
+
+## What is this?
+Life Preserver is a simple project designed to be used with the 
+IBM Cloud Schematics service to make the process of inviting 
+someone to your IBM Cloud account for the purpose of helping you 
+troubleshoot issues with your IBM Cloud Kubernetes Service Clusters
+ (including OpenShift on IBM Cloud clusters).
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:8]:
```suggestion
The Life Preserver project uses IBM Cloud Schematics to 
quickly invite a member of IBM Cloud Support to your IBM 
Cloud account. With this temporary access, IBM Cloud Support 
can help you troubleshoot issues with your IBM Cloud Kubernetes 
Service and Red Hat OpenShift on IBM Cloud clusters.
```

File: iks-life-preserver/README.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+> Note: In terms of Kubernetes RBAC, this user will not have access to any Secrets or RBAC resources inside the cluster.
+
+
+### Step 1
+
+ Login to the IBM Cloud web console at https://cloud.ibm.com.
+
+### Step 2 
+
+Browse to the IBM Cloud Schematics offering at https://cloud.ibm.com/schematics
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:25]:
```suggestion
[Open the IBM Cloud Schematics dashboard.](https://cloud.ibm.com/schematics)
```

File: iks-life-preserver/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+### Step 3
+
+Click `Create workspace` and give your new workspace a relevant name and pick a location and click `Create`.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:29]:
```suggestion
Click **Create Workspace**. Enter a name, select a location, and click **Create**.
```

Comment (Rachael-Graham) [reply iks-life-preserver/README.md:29]: does this workspace need to be in the same RG as the cluster?

File: iks-life-preserver/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Browse to the IBM Cloud Schematics offering at https://cloud.ibm.com/schematics
+
+### Step 3
+
+Click `Create workspace` and give your new workspace a relevant name and pick a location and click `Create`. 
+
+### Step 4
+
+Now you should be looking at your workspace settings. Provide `https://github.com/jpapejr/iks-life-preserver/tree/main` as the Terraform template URL and ensure that you choose a Terraform version of 0.12 or higher for this workspace. Click `Save template information` to save your changes. Wait for the system import and process the Terraform data before continuing.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:33]:
```suggestion
The workspace settings pane opens. In the **Import your Terraform template** section, enter `https://github.com/BM-Cloud/kube-samples/iks-life-preserver/tree/main` for the Terraform template **URL**, and select a **Terraform version** of `terraform_v0.12` or later.

Click **Save template information** and wait for Schematics to import and process the Terraform data before you continue.
```

File: iks-life-preserver/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+### Step 5
+
+You should now see four variables that need to be populated before continuing:
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:37]:
```suggestion
In the **Variables** section, enter **Override values** for the following variables:
```

File: iks-life-preserver/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+* `reference` - This is a user-defined reference that will be appeanded to the generated service ID name for purpose reference.
+* `cluster` - This is the **cluster ID** of the specific cluster to grant **read-only** access to.
+* `apikey` - This is an IBM Cloud API Key with sufficient privileges to invite new members to the account. For security purposes, ensure you check the `Sensitive` box on the far right so your key is obscured from view after it input.
+* `account` - This is the target account within which the operations will occur.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:42]: moved some around to match the order in the ui
```suggestion
* `cluster` - The ID of the cluster to grant _read-only_ access to.
* `apikey` - An IBM Cloud API key with [sufficient privileges to invite users to the account](https://cloud.ibm.com/docs/account?topic=account-iamuserinv#invite-access). For security purposes, select the **Sensitive** checkbox to obscure your API key from view after changes are saved.
* `reference` - A reference that is appended to the generated service ID name.
* `account` - The IBM Cloud account ID which the cluster is created in.
```

File: iks-life-preserver/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+### Step 5
+
+You should now see four variables that need to be populated before continuing:
+
+* `reference` - This is a user-defined reference that will be appeanded to the generated service ID name for purpose reference.
+* `cluster` - This is the **cluster ID** of the specific cluster to grant **read-only** access to.
+* `apikey` - This is an IBM Cloud API Key with sufficient privileges to invite new members to the account. For security purposes, ensure you check the `Sensitive` box on the far right so your key is obscured from view after it input.
+* `account` - This is the target account within which the operations will occur. 
+
+When done, click `Save changes`.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:44]:
```suggestion
Click `Save changes`.
```

File: iks-life-preserver/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+### Step 6
+
+At the top of the screen click on the `Generate plan` button. Ensure it completes successfully before moving on. Address any Terraform errors that might arise due to back API Key or cluster ID values.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:48]:
```suggestion
At the top of the settings pane, click **Generate plan**. In the **Activity** pane, monitor events to ensure that the plan generates successfully before you continue. To review any Terraform errors, click **View log** on the `Failed to generate plan` event card. Then, change the variable override values in the settings pane, and generate the plan again.
```

File: iks-life-preserver/README.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+When done, click `Save changes`.
+
+### Step 6
+
+At the top of the screen click on the `Generate plan` button. Ensure it completes successfully before moving on. Address any Terraform errors that might arise due to back API Key or cluster ID values. 
+
+### Step 7
+
+Again, at the top of the screen, click on the `Apply plan` button. Once the `apply` operation has completed, you can click on `View Logs` to find the generated Service ID API Key. Share this securely with the individual providing assistance; they will be able to authenticate using this API Key for the targeted account & cluster resources as defined above.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:52]:
```suggestion
After the plan generates successfully, click **Apply plan**. After the plan applies successfully, click **View Log** and find the generated Service ID API key. Securely share this generated API key with the IBM Cloud Support engineer who is providing assistance. IBM Cloud Support uses this API key to authenticate with the specified account and view the cluster. 
```

File: iks-life-preserver/README.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+At the top of the screen click on the `Generate plan` button. Ensure it completes successfully before moving on. Address any Terraform errors that might arise due to back API Key or cluster ID values. 
+
+### Step 7
+
+Again, at the top of the screen, click on the `Apply plan` button. Once the `apply` operation has completed, you can click on `View Logs` to find the generated Service ID API Key. Share this securely with the individual providing assistance; they will be able to authenticate using this API Key for the targeted account & cluster resources as defined above. 
+
+### Step 8
+
+At this point, if the `apply` operation was successful you should be able to browse into the [IAM portion of the console](https://cloud.ibm.com/iam/users) to confirm the access.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:56]:
```suggestion
Optional: In the [Users dashboard of the IAM console](https://cloud.ibm.com/iam/users), confirm the access of the IBM Cloud Support engineer. 
```

File: iks-life-preserver/README.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+Again, at the top of the screen, click on the `Apply plan` button. Once the `apply` operation has completed, you can click on `View Logs` to find the generated Service ID API Key. Share this securely with the individual providing assistance; they will be able to authenticate using this API Key for the targeted account & cluster resources as defined above. 
+
+### Step 8
+
+At this point, if the `apply` operation was successful you should be able to browse into the [IAM portion of the console](https://cloud.ibm.com/iam/users) to confirm the access. 
+
+
+
+### Step 9 (roll-back/revert the process)
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:60]:
```suggestion
### Step 9 (Revoke access)
```

File: iks-life-preserver/README.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+From the main workspace page, click on the `Actions` drop-down button near the top-right part of the screen and select `Delete`:
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:62]: In the overview for the Schematics workspace, click the **Actions...** drop-down list and select `Delete`.

File: iks-life-preserver/README.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+* If you want to remove all the changes and delete the workspace completely, check both boxes.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:64]:
```suggestion
* To remove all changes and delete the workspace completely, select both checkboxes.
```

File: iks-life-preserver/README.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+* If you want to remove all the changes and keep the workspace around for re-use at a later time, just check the second box.
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:65]:
```suggestion
* To remove all changes and keep the workspace for re-use at a later time, select only the second checkbox. 
```

File: iks-life-preserver/README.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+
+### Step 9 (roll-back/revert the process)
+
+From the main workspace page, click on the `Actions` drop-down button near the top-right part of the screen and select `Delete`:
+
+* If you want to remove all the changes and delete the workspace completely, check both boxes.
+* If you want to remove all the changes and keep the workspace around for re-use at a later time, just check the second box. 
+
+Confirm the operation by typing in the workspace name and then clicking `Delete`
```
Comment (Rachael-Graham) [inline iks-life-preserver/README.md:67]:
```suggestion
Confirm the operation by entering the workspace name and clicking **Delete**.
```

**Review by Rachael-Graham — COMMENTED** [review]
docs review for IBM Cloud style

PS @jpapejr are you alright with me adding a short blurb about this in the docs? Let me know if you have any specific vision for how this should look, otherwise i can draft and send to you

File: iks-life-preserver/main.tf [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  }
+}
+
+data "external" "apikey" {
+  depends_on = [ ibm_iam_service_id.lifeperserver ]
+  program = ["bash", "${path.root}/createKey.sh", var.apikey, var.account, var.reference]
+}
+
+
+# resource "ibm_iam_user_policy" "policy" {
```
Comment (bhpratt) [inline iks-life-preserver/main.tf:26]: If you're going to keep the commented-out section, I'd suggest providing some in-line comment explaining when it might be used. If it isn't needed, then removed.

File: iks-life-preserver/variables.tf [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+    description = "The cluster ID of the target cluster"
+}
+
+variable "apikey" {
+    description = "An IBM Cloud API Key with permission to invite users to the account"
+}
+
+variable "reference" {
+    description = "A reference tag/string of your choosing (I.e case number, etc.)"
+}
```
Comment (bhpratt) [inline iks-life-preserver/variables.tf:11]: maybe make this optional by providing a default value? don't feel strongly - fine as is.

**Review by bhpratt — APPROVED** [review]

File: iks-life-preserver/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpapejr) [reply iks-life-preserver/README.md:8]: @Rachael-Graham once merged, I'm totally fine with inclusion in the IBM Cloud docs site.

File: iks-life-preserver/README.md (hunk 3 above)
Comment (jpapejr) [reply iks-life-preserver/README.md:29]: It does not.

File: iks-life-preserver/variables.tf (hunk 17 above)
Comment (jpapejr) [reply iks-life-preserver/variables.tf:11]: I had some trouble coming up with a meaningful default. `life-preserver` perhaps? I really expect the customer to provide something here like a case number or something like that. If anyone else feels strongly about it, I'll consider it more.

File: iks-life-preserver/main.tf (hunk 16 above)
Comment (jpapejr) [reply iks-life-preserver/main.tf]: Done in https://github.com/IBM-Cloud/kube-samples/pull/94/commits/85c699c6b8bdc2d945b7658b2c63604c40e5fa0d

File: iks-life-preserver/README.md (hunk 12 above)
Comment (jpapejr) [reply iks-life-preserver/README.md]: updated in https://github.com/IBM-Cloud/kube-samples/pull/94/commits/62fa18b4f63dbce7796907aa3d4a193d97776fa9

**Review by Rachael-Graham — APPROVED** [review]
@jpapejr LMK when you are ready for me to merge

**Comment by jpapejr** [comment]
@Rachael-Graham LGTM
